phys/module_ra_eclipse.F

```diff
@@ -0,0 +1,439 @@
+!
+! Besseliam elements from file eclipse_besselian_elements.dat
+!	X, Y         :: coordinates of the eclipse's axis in the fundamental plane
+!       d, mu        :: declination and azimuth (deg)
+!       l1, l2       :: radii of the penumbra and umbra regions in the fundamental 
+!                       plane
+!	tanf1, tanf2 :: angle of the shadow's conus
+!       t0, min, max :: time when the Besselian elements are valid
+!       Dt           :: time correction
+  integer, parameter       :: NBessel = 4 ! Dimension
+  real, dimension(NBessel) :: X, Y, d, l1, l2, mu
+  real                     :: tanf1, tanf2, t0, tmin, tmax, Dt
+!
+! d and mu in radians
+  real                     :: rmu, rd
+!
+! Besselian elements at the current time t1
+  real                 :: cx, cy, cd, cl1, cl2, cmu
+!
+! Current time t1, and hourly angle H
+  real                 :: t1, &
+                           H   
+!
+! Latitude and longitude in radians
+  real                 :: rlat, rlon
+!
+!
+! Geocentric coordinates
+!----------
+  real                 :: cos_lat1, sin_lat1 ! cos/sin of the corrected latitude
+  real                 :: xi, eta1, zeta1      ! Geocentric coordinates at the observation point
+!
+! Eclipse variables
+  real                 :: DELTA, DELTA2      ! Distance of the observation point
+                                             ! to the eclipse axis
+  real                 :: LL1, LL2           ! Penumbra and Umbra radius
+                                             ! from the point observer's plane
+  logical              :: eclipse_exist
+!
+! Loop indeces
+  integer              :: i, j
+!
+! Internal variables
+  real                 :: A, B
+  real                 :: da,eot,xt24
+  integer              :: julday
+!
+! Eclipse path
+  real :: E, E_1, tan_rd1, rd1, rho1, cy1, tan_gamma, cgamma, sin_beta, &
+          tan_C, C, cm, tan_lat1, tan_lat, sin_H,                       &
+          cos_H, tan_H, beta, lon, tan_rd2, rd2, rho2
+!
+! Missing value
+  real :: MISSING
+
+
+!--------------
+!----------
+! CODE CUSTOMIZATION
+!----------
+!--------------
+!
+  MISSING = 1E30
+
+!--------------
+!----------
+! MAIN
+!----------
+!--------------
+!
+!----------
+! Check if the eclipse is enabled by the namelist.input
+! If eclipse is not enabled -> set eclipse variables to zero and nothing to do
+  if(sw_eclipse .eq. 0) then
+     Obscur(:,:) = 0.
+     mask(:,:)   = 0
+     elat_track  = MISSING
+     elon_track  = MISSING
+     return
+  endif
+!
+!----------
+! Set current time
+  xt24   = mod(xtime+radt,1440.)
+  t1     = gmt+xt24/60
+  julday = int(julian)+1
+  if(t1 .ge. 24.) then
+	t1     = t1 - 24.
+  endif 
+!
+!----------
+! Check if the eclipse exists for the current simulation date
+  eclipse_exist = .false.
+!
+! and load the besselian elements
+! Note: eclipse_besselian_elements.dat should be in the running directory!
+  call load_besselian_elements(t1, julday, year, X, Y, d, l1, l2, mu, & 
```

Comment:
@irowebbn 
Isaac,
Does this "read" happen every time in the routine? Can we not read it once and save it? What happens when there are 10k MPI tasks, do they all open the file?